File: swift/Sources/AetherMeshProtocol/DTN/DtnService.swift
// SPDX-License-Identifier: MIT

import Foundation

/// Default DTN service. Three-tier delivery:
///   direct mesh send → DTN epidemic replication → backend relay.
public actor DtnService {
    private let sender: any MeshSender
    private let store: any BundleStore
    private let strategy: any ReplicationStrategy
    private let incentives: any IncentiveProvider
    private let backend: any BackendClient
    private var reputation: NodeReputationService?

    public var onBundleDelivered: (@Sendable (DtnDeliveryReceipt) -> Void)?

    public init(
        sender: any MeshSender,
        store: any BundleStore = InMemoryBundleStore(),
        strategy: any ReplicationStrategy = GeohashEpidemicStrategy(),
        incentives: any IncentiveProvider = NoopIncentiveProvider(),
        backend: any BackendClient = NoopBackendClient()
    ) {
        self.sender = sender
        self.store = store
        self.strategy = strategy
        self.incentives = incentives
        self.backend = backend
    }

    public func setReputation(_ rep: NodeReputationService?) {
        self.reputation = rep
    }

    public func setOnBundleDelivered(_ callback: (@Sendable (DtnDeliveryReceipt) -> Void)?) {
        self.onBundleDelivered = callback
    }

    public func createBundle(
        recipientUhid: String,
        encryptedPayload: Data,
        priority: BundlePriority = .normal,
        recipientLastGeohash: String? = nil
    ) async -> DtnBundle {
        var bundle = DtnBundle(
            senderUhid: sender.localUhid,
            recipientUhid: recipientUhid,
            encryptedPayload: encryptedPayload,
            priority: priority.rawValue,
            status: BundleStatus.pending.rawValue,
            senderGeohash: sender.localGeohash,
            recipientLastGeohash: recipientLastGeohash
        )
        await store.save(bundle)

        if await tryDirectDelivery(bundle) {
            bundle = withStatus(bundle, status: .delivered)
            await store.save(bundle)
        }
        return bundle
    }

    public func handle(_ packet: MeshPacket) async {
        switch packet.type {
        case .dtnBundle: await handleBundle(packet)
        case .dtnCustodyAck: await handleCustodyAck(packet)
        case .dtnDeliveryReceipt: await handleDeliveryReceipt(packet)
        default: break
        }
    }

    public func runDeliveryScan() async {
        let active = await store.getActive()
        if active.isEmpty { return }
        let peers = sender.connectedPeers()
        let localGeohash = sender.localGeohash

        for var bundle in active {
            if bundle.status == BundleStatus.delivered.rawValue || bundle.isExpired { continue }
            if await tryDirectDelivery(bundle) {
                bundle = withStatus(bundle, status: .delivered)
                await store.save(bundle)
                continue
            }
            if peers.isEmpty || bundle.copyCount >= bundle.maxCopies { continue }
            let targets = strategy.selectTargets(bundle: bundle, peers: peers, localGeohash: localGeohash)
            for target in targets {
                if bundle.copyCount >= bundle.maxCopies { break }
                let pkt = bundlePacket(bundle)
                if await sender.send(pkt, nextHopUhid: target) {
                    bundle = withCopyCount(bundle, copyCount: bundle.copyCount + 1)
                    await store.save(bundle)
                    await incentives.recordRelay(localUhid: sender.localUhid, packet: pkt)
                }
            }
        }
    }

    public func expireStale() async -> Int { await store.expireStale() }
    public func getActiveBundles() async -> [DtnBundle] { await store.getActive() }

    private func tryDirectDelivery(_ bundle: DtnBundle) async -> Bool {
        let pkt = bundlePacket(bundle)
        for peer in sender.connectedPeers() where peer.uhid == bundle.recipientUhid {
            if await sender.send(pkt, nextHopUhid: bundle.recipientUhid) { return true }
            break
        }
        return await backend.syncDtnBundle(bundle)
    }

    private func bundlePacket(_ bundle: DtnBundle) -> MeshPacket {
        MeshPacket(
            id: bundle.id,
            type: .dtnBundle,
            sourceUhid: sender.localUhid,
            destinationUhid: bundle.recipientUhid,
            ttl: 30,
            priority: UInt8(clamping: bundle.priority),
            payload: encodeBundle(bundle)
        )
    }

    private func handleBundle(_ packet: MeshPacket) async {
        guard let bundle = decodeBundle(packet.payload) else { return }
        if bundle.recipientUhid == sender.localUhid {
            let delivered = withStatus(bundle, status: .delivered)
            await store.save(delivered)
            await sendDeliveryReceipt(delivered)
            await reputation?.recordDeliverySuccess(uhid: packet.sourceUhid, roundTripMs: 0)
            return
        }
        if await store.getActiveCount() >= ProtocolConstants.dtnMaxBundlesPerNode {
            await sendCustodyAck(bundleId: bundle.id, toUhid: packet.sourceUhid, accepted: false)
            return
        }
        let inCustody = withStatus(withHopCount(bundle, hopCount: bundle.hopCount + 1), status: .inCustody)
        await store.save(inCustody)
        await store.saveCustody(CustodyRecord(
            bundleId: bundle.id,
            fromUhid: packet.sourceUhid,
            toUhid: sender.localUhid,
            accepted: true
        ))
        await sendCustodyAck(bundleId: bundle.id, toUhid: packet.sourceUhid, accepted: true)
        await incentives.recordRelay(localUhid: sender.localUhid, packet: packet)
    }

    private func handleCustodyAck(_ packet: MeshPacket) async {
        guard let (bundleId, accepted) = parseCustodyAck(packet.payload) else { return }
        if !accepted {
            await reputation?.recordCustodyRefusal(uhid: packet.sourceUhid)
            return
        }
        guard let b = await store.get(bundleId) else { return }
        await store.save(withCopyCount(b, copyCount: b.copyCount + 1))
    }

    private func handleDeliveryReceipt(_ packet: MeshPacket) async {
        guard let receipt = parseDeliveryReceipt(packet.payload) else { return }
        if let b = await store.get(receipt.bundleId) {
            await store.save(withStatus(b, status: .delivered))
        }
        if let cb = onBundleDelivered { cb(receipt) }
    }

    private func sendCustodyAck(bundleId: UUID, toUhid: String, accepted: Bool) async {
        if toUhid.isEmpty { return }
        let payload = encodeCustodyAck(bundleId: bundleId, accepted: accepted)
        let pkt = MeshPacket(
            type: .dtnCustodyAck,
            sourceUhid: sender.localUhid,
            destinationUhid: toUhid,
            ttl: ProtocolConstants.defaultTtl,
            payload: payload
        )
        _ = await sender.send(pkt, nextHopUhid: toUhid)
    }

    private func sendDeliveryReceipt(_ bundle: DtnBundle) async {
        if bundle.senderUhid.isEmpty || bundle.senderUhid == sender.localUhid { return }
        let custody = await store.getCustodyRecords(bundle.id)
        let receipt = DtnDeliveryReceipt(
            bundleId: bundle.id,
            recipientUhid: bundle.recipientUhid,
            totalHops: bundle.hopCount,
            totalCustodyTransfers: Int32(custody.count),
            deliveredAt: Date()
        )
        let pkt = MeshPacket(
            type: .dtnDeliveryReceipt,
            sourceUhid: sender.localUhid,
            destinationUhid: bundle.senderUhid,
            ttl: ProtocolConstants.defaultTtl,
            payload: encodeDeliveryReceipt(receipt)
        )
        _ = await sender.send(pkt, nextHopUhid: bundle.senderUhid)
    }

    // ─── DtnBundle is immutable; copy with field changes ───
    private func withStatus(_ b: DtnBundle, status: BundleStatus) -> DtnBundle {
        DtnBundle(
            id: b.id, senderUhid: b.senderUhid, recipientUhid: b.recipientUhid,
            encryptedPayload: b.encryptedPayload, priority: b.priority,
            status: status.rawValue, copyCount: b.copyCount, maxCopies: b.maxCopies,
            senderGeohash: b.senderGeohash, recipientLastGeohash: b.recipientLastGeohash,
            hopCount: b.hopCount, createdAt: b.createdAt, expiresAt: b.expiresAt
        )
    }

    private func withCopyCount(_ b: DtnBundle, copyCount: Int32) -> DtnBundle {
        DtnBundle(
            id: b.id, senderUhid: b.senderUhid, recipientUhid: b.recipientUhid,
            encryptedPayload: b.encryptedPayload, priority: b.priority,
            status: b.status, copyCount: copyCount, maxCopies: b.maxCopies,
            senderGeohash: b.senderGeohash, recipientLastGeohash: b.recipientLastGeohash,
            hopCount: b.hopCount, createdAt: b.createdAt, expiresAt: b.expiresAt
        )
    }

    private func withHopCount(_ b: DtnBundle, hopCount: Int32) -> DtnBundle {
        DtnBundle(
            id: b.id, senderUhid: b.senderUhid, recipientUhid: b.recipientUhid,
            encryptedPayload: b.encryptedPayload, priority: b.priority,
            status: b.status, copyCount: b.copyCount, maxCopies: b.maxCopies,
            senderGeohash: b.senderGeohash, recipientLastGeohash: b.recipientLastGeohash,
            hopCount: hopCount, createdAt: b.createdAt, expiresAt: b.expiresAt
        )
    }
}

// ─── snake_case JSON wire helpers ───

private struct BundleWire: Codable {
    let id: UUID
    let sender_uhid: String
    let recipient_uhid: String
    let encrypted_payload: Data
    let priority: Int32
    let status: Int32
    let copy_count: Int32
    let max_copies: Int32
    let sender_geohash: String?
    let recipient_last_geohash: String?
    let hop_count: Int32
    let created_at_ms: Int64
    let expires_at_ms: Int64
}

private struct CustodyAckWire: Codable {
    let bundle_id: UUID
    let accepted: Bool
}

private struct DeliveryReceiptWire: Codable {
    let bundle_id: UUID
    let recipient_uhid: String
    let total_hops: Int32
    let total_custody_transfers: Int32
    let delivered_at_ms: Int64
}

private func encodeBundle(_ b: DtnBundle) -> Data {
    let wire = BundleWire(
        id: b.id,
        sender_uhid: b.senderUhid,
        recipient_uhid: b.recipientUhid,
        encrypted_payload: b.encryptedPayload,
        priority: b.priority,
        status: b.status,
        copy_count: b.copyCount,
        max_copies: b.maxCopies,
        sender_geohash: b.senderGeohash,
        recipient_last_geohash: b.recipientLastGeohash,
        hop_count: b.hopCount,
        created_at_ms: Int64(b.createdAt.timeIntervalSince1970 * 1000),
        expires_at_ms: Int64(b.expiresAt.timeIntervalSince1970 * 1000)
    )
    return (try? JSONEncoder().encode(wire)) ?? Data()
}

private func decodeBundle(_ data: Data) -> DtnBundle? {
    guard let wire = try? JSONDecoder().decode(BundleWire.self, from: data) else { return nil }
    return DtnBundle(
        id: wire.id,
        senderUhid: wire.sender_uhid,
        recipientUhid: wire.recipient_uhid,
        encryptedPayload: wire.encrypted_payload,
        priority: wire.priority,
        status: wire.status,
        copyCount: wire.copy_count,
        maxCopies: wire.max_copies,
        senderGeohash: wire.sender_geohash,
        recipientLastGeohash: wire.recipient_last_geohash,
        hopCount: wire.hop_count,
        createdAt: Date(timeIntervalSince1970: TimeInterval(wire.created_at_ms) / 1000),
        expiresAt: Date(timeIntervalSince1970: TimeInterval(wire.expires_at_ms) / 1000)
    )
}

private func encodeCustodyAck(bundleId: UUID, accepted: Bool) -> Data {
    let w = CustodyAckWire(bundle_id: bundleId, accepted: accepted)
    return (try? JSONEncoder().encode(w)) ?? Data()
}

private func parseCustodyAck(_ data: Data) -> (UUID, Bool)? {
    guard let w = try? JSONDecoder().decode(CustodyAckWire.self, from: data) else { return nil }
    return (w.bundle_id, w.accepted)
}

private func encodeDeliveryReceipt(_ r: DtnDeliveryReceipt) -> Data {
    let w = DeliveryReceiptWire(
        bundle_id: r.bundleId,
        recipient_uhid: r.recipientUhid,
        total_hops: r.totalHops,
        total_custody_transfers: r.totalCustodyTransfers,
        delivered_at_ms: Int64(r.deliveredAt.timeIntervalSince1970 * 1000)
    )
    return (try? JSONEncoder().encode(w)) ?? Data()
}

private func parseDeliveryReceipt(_ data: Data) -> DtnDeliveryReceipt? {
    guard let w = try? JSONDecoder().decode(DeliveryReceiptWire.self, from: data) else { return nil }
    return DtnDeliveryReceipt(
        bundleId: w.bundle_id,
        recipientUhid: w.recipient_uhid,
        totalHops: w.total_hops,
        totalCustodyTransfers: w.total_custody_transfers,
        deliveredAt: Date(timeIntervalSince1970: TimeInterval(w.delivered_at_ms) / 1000)
    )
}

File: src/AetherMesh.Core/Dtn/DtnService.cs
// SPDX-License-Identifier: MIT

using System.Text.Json;
using AetherMesh.Constants;
using AetherMesh.Diagnostics;
using AetherMesh.Extensibility;
using AetherMesh.Models;
using AetherMesh.Protocol;
using AetherMesh.Reputation;
using AetherMesh.Routing;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace AetherMesh.Dtn;

/// <summary>
/// Default DTN service implementation. Bundles are JSON-serialized into <see cref="MeshPacket.Payload"/>
/// using <see cref="JsonNamingPolicy.SnakeCaseLower"/> for cross-language interoperability.
/// </summary>
public sealed class DtnService : IDtnService
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
    };

    private readonly IMeshSender _sender;
    private readonly IDtnBundleStore _store;
    private readonly IBundleReplicationStrategy _strategy;
    private readonly IAetherMeshIncentiveProvider _incentives;
    private readonly IAetherMeshBackendClient _backend;
    private readonly INodeReputationService? _reputation;
    private readonly ILogger<DtnService> _logger;

    public event EventHandler<DtnDeliveryReceipt>? BundleDelivered;

    public DtnService(
        IMeshSender sender,
        IDtnBundleStore? store = null,
        IBundleReplicationStrategy? strategy = null,
        IAetherMeshIncentiveProvider? incentives = null,
        IAetherMeshBackendClient? backend = null,
        INodeReputationService? reputation = null,
        ILogger<DtnService>? logger = null)
    {
        _sender = sender ?? throw new ArgumentNullException(nameof(sender));
        _store = store ?? new InMemoryDtnBundleStore();
        _strategy = strategy ?? new GeohashEpidemicStrategy();
        _incentives = incentives ?? new DefaultIncentiveProvider();
        _backend = backend ?? new DefaultBackendClient();
        _reputation = reputation;
        _logger = logger ?? NullLogger<DtnService>.Instance;
    }

    public async Task<DtnBundle> CreateBundleAsync(
        string recipientUhid,
        byte[] encryptedPayload,
        BundlePriority priority = BundlePriority.Normal,
        string? recipientLastGeohash = null,
        CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(recipientUhid);
        ArgumentNullException.ThrowIfNull(encryptedPayload);

        var bundle = new DtnBundle
        {
            SenderUhid = _sender.LocalUhid,
            RecipientUhid = recipientUhid,
            EncryptedPayload = encryptedPayload,
            Priority = priority,
            MaxCopies = ProtocolConstants.DtnMaxCopies,
            SenderGeohash = _sender.LocalGeohash,
            RecipientLastGeohash = recipientLastGeohash,
            ExpiresAt = DateTime.UtcNow.AddHours(ProtocolConstants.DtnBundleTtlHours),
        };

        await _store.SaveAsync(bundle, cancellationToken).ConfigureAwait(false);
        _logger.LogDebug("DTN bundle {Id} created for {Recipient} priority={Priority}",
            bundle.Id, recipientUhid, priority);

        if (await TryDirectDeliveryAsync(bundle, cancellationToken).ConfigureAwait(false))
        {
            bundle.Status = BundleStatus.Delivered;
            await _store.SaveAsync(bundle, cancellationToken).ConfigureAwait(false);
            return bundle;
        }

        return bundle;
    }

    public async Task HandleAsync(MeshPacket packet, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(packet);

        switch (packet.Type)
        {
            case PacketType.DtnBundle:
                await HandleBundleAsync(packet, cancellationToken).ConfigureAwait(false);
                break;
            case PacketType.DtnCustodyAck:
                await HandleCustodyAckAsync(packet, cancellationToken).ConfigureAwait(false);
                break;
            case PacketType.DtnDeliveryReceipt:
                await HandleDeliveryReceiptAsync(packet, cancellationToken).ConfigureAwait(false);
                break;
            default:
                _logger.LogDebug("DTN HandleAsync ignoring non-DTN packet type {Type}", packet.Type);
                break;
        }
    }

    public async Task RunDeliveryScanAsync(CancellationToken cancellationToken = default)
    {
        var active = await _store.GetActiveAsync(cancellationToken).ConfigureAwait(false);
        if (active.Count == 0) return;

        var peers = _sender.GetConnectedPeers();
        _logger.LogDebug("DTN delivery scan — {Count} active bundles, {Peers} peers", active.Count, peers.Count);

        foreach (var bundle in active)
        {
            if (cancellationToken.IsCancellationRequested) break;
            if (bundle.Status == BundleStatus.Delivered || bundle.IsExpired) continue;

            if (await TryDirectDeliveryAsync(bundle, cancellationToken).ConfigureAwait(false))
            {
                bundle.Status = BundleStatus.Delivered;
                await _store.SaveAsync(bundle, cancellationToken).ConfigureAwait(false);
                continue;
            }

            if (peers.Count == 0 || bundle.CopyCount >= bundle.MaxCopies) continue;

            var targets = _strategy.SelectTargets(bundle, peers, _sender.LocalGeohash);
            foreach (var target in targets)
            {
                if (cancellationToken.IsCancellationRequested) break;
                if (bundle.CopyCount >= bundle.MaxCopies) break;

                var packet = BuildBundlePacket(bundle, target);
                if (await _sender.SendAsync(packet, target, cancellationToken).ConfigureAwait(false))
                {
                    bundle.CopyCount++;
                    await _store.SaveAsync(bundle, cancellationToken).ConfigureAwait(false);
                    await _incentives.RecordRelayAsync(_sender.LocalUhid, packet, cancellationToken).ConfigureAwait(false);
                    _logger.LogDebug("DTN bundle {Id} replicated to {Target} copyCount={Copies}",
                        bundle.Id, target, bundle.CopyCount);
                }
            }
        }
    }

    public async Task<int> ExpireStaleAsync(CancellationToken cancellationToken = default)
    {
        var expired = await _store.ExpireStaleAsync(cancellationToken).ConfigureAwait(false);
        if (expired > 0)
            AetherMeshTelemetry.DtnBundlesExpired.Add(expired);
        return expired;
    }

    public Task<IReadOnlyList<DtnBundle>> GetActiveBundlesAsync(CancellationToken cancellationToken = default)
        => _store.GetActiveAsync(cancellationToken);

    private async Task<bool> TryDirectDeliveryAsync(DtnBundle bundle, CancellationToken cancellationToken)
    {
        var packet = BuildBundlePacket(bundle, bundle.RecipientUhid);

        var directPeer = _sender.GetConnectedPeers()
            .FirstOrDefault(p => string.Equals(p.Uhid, bundle.RecipientUhid, StringComparison.Ordinal));
        if (directPeer is not null)
        {
            if (await _sender.SendAsync(packet, bundle.RecipientUhid, cancellationToken).ConfigureAwait(false))
            {
                _logger.LogDebug("DTN bundle {Id} delivered directly to {Recipient}", bundle.Id, bundle.RecipientUhid);
                return true;
            }
        }

        var relayed = await _backend.SyncDtnBundleAsync(bundle, cancellationToken).ConfigureAwait(false);
        if (relayed)
        {
            _logger.LogDebug("DTN bundle {Id} accepted by backend relay", bundle.Id);
            return true;
        }

        return false;
    }

    private async Task HandleBundleAsync(MeshPacket packet, CancellationToken cancellationToken)
    {
        DtnBundle? bundle;
        try
        {
            bundle = JsonSerializer.Deserialize<DtnBundle>(packet.Payload, JsonOptions);
        }
        catch (JsonException ex)
        {
            _logger.LogWarning(ex, "DTN: failed to deserialize bundle from packet {Id}", packet.Id);
            return;
        }
        if (bundle is null) return;

        if (string.Equals(bundle.RecipientUhid, _sender.LocalUhid, StringComparison.Ordinal))
        {
            bundle.Status = BundleStatus.Delivered;
            await _store.SaveAsync(bundle, cancellationToken).ConfigureAwait(false);
            AetherMeshTelemetry.DtnBundlesDelivered.Add(1);
            _logger.LogDebug("DTN bundle {Id} delivered locally — {Hops} hops", bundle.Id, bundle.HopCount);
            _ = _reputation?.RecordDeliverySuccessAsync(packet.SourceUhid, 0);
            await SendDeliveryReceiptAsync(bundle, cancellationToken).ConfigureAwait(false);
            return;
        }

        var activeCount = await _store.GetActiveCountAsync(cancellationToken).ConfigureAwait(false);
        if (activeCount >= ProtocolConstants.DtnMaxBundlesPerNode)
        {
            await SendCustodyAckAsync(bundle.Id, packet.SourceUhid, accepted: false, cancellationToken).ConfigureAwait(false);
            _logger.LogDebug("DTN custody refused for {Id} — at capacity ({Count})", bundle.Id, activeCount);
            return;
        }

        using var custodyActivity = AetherMeshTelemetry.ActivitySource.StartActivity("AetherMesh.Dtn.Custody");
        if (custodyActivity is not null)
        {
            custodyActivity.SetTag("aethermesh.bundle.id", bundle.Id);
            custodyActivity.SetTag("aethermesh.recipient.uhid", AetherMeshTelemetry.SanitizeUhid(bundle.RecipientUhid));
        }

        bundle.Status = BundleStatus.InCustody;
        bundle.HopCount++;
        await _store.SaveAsync(bundle, cancellationToken).ConfigureAwait(false);
        await _store.SaveCustodyAsync(new CustodyRecord
        {
            BundleId = bundle.Id,
            FromUhid = packet.SourceUhid,
            ToUhid = _sender.LocalUhid,
            Accepted = true,
        }, cancellationToken).ConfigureAwait(false);

        await SendCustodyAckAsync(bundle.Id, packet.SourceUhid, accepted: true, cancellationToken).ConfigureAwait(false);
        await _incentives.RecordRelayAsync(_sender.LocalUhid, packet, cancellationToken).ConfigureAwait(false);

        AetherMeshTelemetry.DtnBundlesAccepted.Add(1);
        _logger.LogDebug("DTN custody accepted for {Id} from {From} hops={Hops}",
            bundle.Id, packet.SourceUhid, bundle.HopCount);
    }

    private async Task HandleCustodyAckAsync(MeshPacket packet, CancellationToken cancellationToken)
    {
        CustodyAckPayload? ack;
        try
        {
            ack = JsonSerializer.Deserialize<CustodyAckPayload>(packet.Payload, JsonOptions);
        }
        catch (JsonException ex)
        {
            _logger.LogWarning(ex, "DTN: failed to deserialize custody ack from packet {Id}", packet.Id);
            return;
        }
        if (ack is null || ack.BundleId == Guid.Empty) return;

        if (!ack.Accepted)
        {
            _logger.LogDebug("DTN custody ack: {Receiver} refused custody of {Bundle}", packet.SourceUhid, ack.BundleId);
            _ = _reputation?.RecordCustodyRefusalAsync(packet.SourceUhid);
            return;
        }

        var bundle = await _store.GetAsync(ack.BundleId, cancellationToken).ConfigureAwait(false);
        if (bundle is null) return;
        bundle.CopyCount++;
        await _store.SaveAsync(bundle, cancellationToken).ConfigureAwait(false);
        _logger.LogDebug("DTN custody ack — bundle {Id} copies now {Count}", bundle.Id, bundle.CopyCount);
    }

    private async Task HandleDeliveryReceiptAsync(MeshPacket packet, CancellationToken cancellationToken)
    {
        DtnDeliveryReceipt? receipt;
        try
        {
            receipt = JsonSerializer.Deserialize<DtnDeliveryReceipt>(packet.Payload, JsonOptions);
        }
        catch (JsonException ex)
        {
            _logger.LogWarning(ex, "DTN: failed to deserialize delivery receipt from packet {Id}", packet.Id);
            return;
        }
        if (receipt is null || receipt.BundleId == Guid.Empty) return;

        var bundle = await _store.GetAsync(receipt.BundleId, cancellationToken).ConfigureAwait(false);
        if (bundle is not null)
        {
            bundle.Status = BundleStatus.Delivered;
            await _store.SaveAsync(bundle, cancellationToken).ConfigureAwait(false);
            AetherMeshTelemetry.DtnBundlesDelivered.Add(1);
        }

        _logger.LogDebug("DTN delivery receipt — bundle {Id} delivered after {Hops} hops, {Transfers} custody transfers",
            receipt.BundleId, receipt.TotalHops, receipt.TotalCustodyTransfers);
        BundleDelivered?.Invoke(this, receipt);
    }

    private MeshPacket BuildBundlePacket(DtnBundle bundle, string nextHopUhid)
    {
        return new MeshPacket
        {
            Type = PacketType.DtnBundle,
            SourceUhid = _sender.LocalUhid,
            DestinationUhid = bundle.RecipientUhid,
            Ttl = ProtocolConstants.DtnTtl,
            Priority = (byte)Math.Clamp((int)bundle.Priority, 0, byte.MaxValue),
            Payload = JsonSerializer.SerializeToUtf8Bytes(bundle, JsonOptions),
        };
    }

    private async Task SendCustodyAckAsync(Guid bundleId, string toUhid, bool accepted, CancellationToken cancellationToken)
    {
        if (string.IsNullOrEmpty(toUhid)) return;

        var payload = JsonSerializer.SerializeToUtf8Bytes(new CustodyAckPayload
        {
            BundleId = bundleId,
            Accepted = accepted,
        }, JsonOptions);

        var packet = new MeshPacket
        {
            Type = PacketType.DtnCustodyAck,
            SourceUhid = _sender.LocalUhid,
            DestinationUhid = toUhid,
            Ttl = ProtocolConstants.DefaultTtl,
            Payload = payload,
        };
        await _sender.SendAsync(packet, toUhid, cancellationToken).ConfigureAwait(false);
    }

    private async Task SendDeliveryReceiptAsync(DtnBundle bundle, CancellationToken cancellationToken)
    {
        if (string.IsNullOrEmpty(bundle.SenderUhid) || string.Equals(bundle.SenderUhid, _sender.LocalUhid, StringComparison.Ordinal))
            return;

        var custody = await _store.GetCustodyRecordsAsync(bundle.Id, cancellationToken).ConfigureAwait(false);

        var receipt = new DtnDeliveryReceipt
        {
            BundleId = bundle.Id,
            RecipientUhid = bundle.RecipientUhid,
            TotalHops = bundle.HopCount,
            TotalCustodyTransfers = custody.Count,
        };

        var packet = new MeshPacket
        {
            Type = PacketType.DtnDeliveryReceipt,
            SourceUhid = _sender.LocalUhid,
            DestinationUhid = bundle.SenderUhid,
            Ttl = ProtocolConstants.DefaultTtl,
            Payload = JsonSerializer.SerializeToUtf8Bytes(receipt, JsonOptions),
        };

        await _sender.SendAsync(packet, bundle.SenderUhid, cancellationToken).ConfigureAwait(false);
    }

    private sealed class CustodyAckPayload
    {
        public Guid BundleId { get; set; }
        public bool Accepted { get; set; }
    }

    private sealed class DefaultIncentiveProvider : IAetherMeshIncentiveProvider
    {
    }

    private sealed class DefaultBackendClient : IAetherMeshBackendClient
    {
    }
}
